network/dag/bbolt_tree.go

```diff
@@ -1,38 +1,143 @@
+func (store *bboltTree) dagObserver(ctx context.Context, transaction Transaction, _ []byte) {
+	if transaction != nil { // can happen when payload is written for private TX
+		err := storage.BBoltTXUpdate(ctx, store.db, func(callbackCtx context.Context, tx *bbolt.Tx) error {
+			dirty := store.tree.InsertGetDirty(transaction.Ref(), transaction.Clock())
+
+			// Rollback after timeout to bring tree and DAG back in sync.
+			// A call to writeUpdates will persist all uncommitted tree changes. So a failed bboltTx will be dropped by the dag and (eventually) persisted by the tree.
+			c, cancel := context.WithTimeout(context.Background(), observerRollbackTimeOut) // << timeout must not be shorter than expected write operation to disk
+			go func() {
+				<-c.Done()
+				err := c.Err()
+				if err == context.DeadlineExceeded {
+					log.Logger().Warnf("deadline exceeded - rollback transaction %s from %s", transaction.Ref(), store.bucketName)
+					store.tree.Delete(transaction.Ref(), transaction.Clock())
+				}
+			}()
+			tx.OnCommit(func() {
+				store.tree.ResetUpdate()
+				cancel()
+			})
+
+			return store.writeUpdates(callbackCtx, dirty, nil)
+		})
+		if err != nil {
+			log.Logger().Errorf("failed to add transaction to %s: %s", store.bucketName, err.Error())
+		}
+	}
+}
+
+// migrate builds a tree by walking over the dag and adding all Transaction references to the tree without checking for validity.
+// The tree is stored on disk once it is in sync with the dag.
+// Should only be called on an empty tree.
+func (store *bboltTree) migrate(ctx context.Context, state State) error {
+	if !store.isEmpty() {
+		return nil
+	}
+
+	err := state.Walk(ctx, func(_ context.Context, transaction Transaction) bool {
+		store.tree.Insert(transaction.Ref(), transaction.Clock())
+		return true
+	}, hash.EmptyHash())
+	if err != nil {
+		return err
+	}
+
+	dirties, orphaned := store.tree.GetUpdates()
+	store.tree.ResetUpdate()
+
+	err = store.writeUpdates(ctx, dirties, orphaned)
+	if err != nil {
+		return err
+	}
+
+	return nil
 }
 
 // read fills the tree with data in the bucket.
-// Returns an error the bucket does not exist, or if data in the bucket doesn't match the prototype.
-func (store bboltTree) read(bucketName string, prototype tree.Data) (tree.Tree, error) {
-	tr := tree.New(prototype.New(), 0)
-	return tr, store.db.View(func(tx *bbolt.Tx) error {
+// Returns an error the bucket does not exist, or if data in the bucket doesn't match the tree's Data prototype.
+func (store *bboltTree) read(ctx context.Context) error {
+	return storage.BBoltTXUpdate(ctx, store.db, func(_ context.Context, tx *bbolt.Tx) error {
 		// get bucket
-		bucket := tx.Bucket([]byte(bucketName))
+		bucket := tx.Bucket([]byte(store.bucketName))
 		if bucket == nil {
-			return fmt.Errorf("bucket '%s' not found", bucketName)
+			// should only happen once for a new tree/bucket.
+			log.Logger().Warnf("tree bucket '%s' does not exist", store.bucketName)
```

Comment:
how is this log useful?

Reply:
the bucket should always exist, except during migration. Maybe change this back to return an error in V3?